core/src/main/java/org/elasticsearch/index/translog/Translog.java

```diff
@@ -1375,54 +1397,85 @@ public void rollGeneration() throws IOException {
         try (ReleasableLock ignored = writeLock.acquire()) {
             ensureOpen();
+            if (committedGeneration > current.generation) {
+                final String message = String.format(
+                        Locale.ROOT,
+                        "tried to commit generation [%d] later than the current generation [%d]",
+                        committedGeneration,
+                        current.generation);
+                throw new IllegalStateException(message);
+            }
+            final Long min = readers.stream().map(TranslogReader::getGeneration).min(Long::compareTo).orElse(Long.MIN_VALUE);
+            if (committedGeneration < min) {
+                final String message = String.format(
+                        Locale.ROOT,
+                        "tried to commit generation [%d] before minimum generation [%d]",
+                        committedGeneration,
+                        min);
+                throw new IllegalStateException(message);
+            }
             if (currentCommittingGeneration == NOT_SET_GENERATION) {
                 prepareCommit();
             }
             assert currentCommittingGeneration != NOT_SET_GENERATION;
             assert readers.stream().anyMatch(r -> r.getGeneration() == currentCommittingGeneration)
                     : "readers missing committing generation [" + currentCommittingGeneration + "]";
             // set the last committed generation otherwise old files will not be cleaned up
-            lastCommittedTranslogFileGeneration = currentCommittingGeneration + 1;
+            lastCommittedTranslogFileGeneration = committedGeneration;
             currentCommittingGeneration = NOT_SET_GENERATION;
             trimUnreferencedReaders();
         }
-        return 0;
     }
 
+    /**
+     * Trims unreferenced translog generations. The guarantee here is that translog generations will be preserved for all outstanding views
+     * and from the last committed translog generation defined by {@link Translog#lastCommittedTranslogFileGeneration}.
+     */
     void trimUnreferencedReaders() {
         try (ReleasableLock ignored = writeLock.acquire()) {
             if (closed.get()) {
-                // we're shutdown potentially on some tragic event - don't delete anything
+                // we're shutdown potentially on some tragic event, don't delete anything
                 return;
             }
-            long minReferencedGen = outstandingViews.stream().mapToLong(View::minTranslogGeneration).min().orElse(Long.MAX_VALUE);
-            minReferencedGen = Math.min(lastCommittedTranslogFileGeneration, minReferencedGen);
-            final long finalMinReferencedGen = minReferencedGen;
-            List<TranslogReader> unreferenced = readers.stream().filter(r -> r.getGeneration() < finalMinReferencedGen).collect(Collectors.toList());
+            long minReferencedGen = Math.min(
```

Comment:
I see a small improvement here which is removing the need for the extra local variable `finalMinReferencedGen` and so avoiding overwriting `minReferencedGen`. I find it easier to read this way.